### DISCUSSION TITLE
Multi users behavior

### DISCUSSION THREAD

**Discussion by jrobinAV** [Avaiga/taipy#72] [Q&A] [answered]
when multi users modify the same data simultaneously, what will happen?

**Comment by FlorianJacta** [comment, marked as answer]
Two answers can be given depending if we are talking about Taipy Core or Taipy GUI.

**Taipy GUI**
In Taipy GUI, each session/user has its own _state_ ([doc](https://docs.taipy.io/en/latest/manuals/reference/taipy.gui.State/)). In this _state_, all the variables present on the GUI are accessible. Changing a data in a _state_ will not change this variable in another _state_ of another session.

**Taipy Core**

In Taipy Core, the concurrency is handled by Taipy directly through the Data Nodes and the Scheduler. When the first user writes a Data Node, Taipy locks the Data Node until the writing is done. If a second user tries to write the same Data Node while it is locked, the writing is rejected. Note that the locked status of the Data Node is exposed to the users so they can know if the Data Node is locked or not.
